### PR TITLE
[3453] Memoizing edges and nodes style

### LINKED ISSUE #3453: Fix unnecessary rerender of edges an nodes
image
This is a trace after dragging a node. We can see all edges were re-rendered and the average time to re-render is 8ms.

The PR associated to this ticket aim to stop re-rendering all edges and to improve average re-render time by memoizing edge et node styles.

### PR DESCRIPTION
**Pull request by mcharfadi**

### CODE REVIEW

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/connector/useConnectorNodeStyle.tsx [hunk 1]
```diff
@@ -43,5 +43,11 @@ export const useConnectorNodeStyle = (nodeId: string, descriptionId: string): Us
...
     }
   }
-  return { style };
+
+  const memoizedStyle = useMemo(
+    () => style,
+    [candidates.map((candidate) => candidate.id).join('-'), isNewConnection, hoveredNode?.id]
+  );
+
+  return { style: memoizedStyle };
```
Comment (frouene) [inline packages/diagrams/frontend/sirius-components-diagrams/src/renderer/connector/useConnectorNodeStyle.tsx:52]: typo `memoized` -> `memorized`

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/dropNode/useDropDiagramStyle.tsx [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+  const targetNode = getNodes().find((node) => node.data.isDropNodeTarget);
+  const draggedNode = getNodes().find((node) => node.dragging);
+
+  const diagramTargeted: boolean = !targetNode && !!draggedNode?.parentNode;
+  const diagramForbidden: boolean = diagramTargeted && !!draggedNode?.id !== null && !droppableOnDiagram;
+  const backgroundColor = diagramForbidden ? theme.palette.action.disabledBackground : theme.palette.background.default;
+
+  const memoizedStyle = useMemo(() => {
+    return {
```
Comment (frouene) [inline packages/diagrams/frontend/sirius-components-diagrams/src/renderer/dropNode/useDropDiagramStyle.tsx:35]: typo `memoized` -> `memorized`

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/dropNode/useDropNodeStyle.ts [hunk 3]
```diff
@@ -41,5 +41,7 @@ export const useDropNodeStyle = (
     }
   }
 
-  return { style };
+  const memoizedStyle = useMemo(() => style, [isDropNodeTarget, isDropNodeCandidate, isDragging, draggedNodeId]);
+
```
Comment (frouene) [inline packages/diagrams/frontend/sirius-components-diagrams/src/renderer/dropNode/useDropNodeStyle.ts:45]: typo `memoized` -> `memorized`

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/node/FreeFormNode.tsx [hunk 4]
```diff
@@ -77,15 +77,20 @@ export const FreeFormNode = memo(({ data, id, selected, dragging }: NodeProps<Fr
 
   const theme = useTheme();
   const { onDrop, onDragOver } = useDrop();
-  const { style: connectionFeedbackStyle } = useConnectorNodeStyle(id, data.nodeDescription.id);
-  const { style: dropFeedbackStyle } = useDropNodeStyle(data.isDropNodeTarget, data.isDropNodeCandidate, dragging);
 
-  const rotation = computeBorderRotation(data);
+  const rotation = useMemo(() => computeBorderRotation(data), [data.isBorderNode, data.positionDependentRotation]);
   let imageURL: string | undefined = undefined;
```
Comment (frouene) [inline packages/diagrams/frontend/sirius-components-diagrams/src/renderer/node/FreeFormNode.tsx:85]: Should depend on `data.borderNodePosition` ?

**Review by frouene — CHANGES_REQUESTED** [review]

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/connector/useConnectorNodeStyle.tsx (hunk 1 above)
Comment (sbegaudeau) [reply packages/diagrams/frontend/sirius-components-diagrams/src/renderer/connector/useConnectorNodeStyle.tsx:52]: Memoizing and memorizing are two different things, I think that Michael is correct here in mentioning memoizing. See the official documentation using this term: https://react.dev/reference/react/useMemo#memoizing-a-dependency-of-another-hook

Comment (sbegaudeau) [reply packages/diagrams/frontend/sirius-components-diagrams/src/renderer/connector/useConnectorNodeStyle.tsx:52]: Or wikipedia: https://en.wikipedia.org/wiki/Memoization :)

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/dropNode/useDropDiagramStyle.tsx (hunk 2 above)
Comment (sbegaudeau) [reply packages/diagrams/frontend/sirius-components-diagrams/src/renderer/dropNode/useDropDiagramStyle.tsx:35]: Same comment

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/dropNode/useDropNodeStyle.ts (hunk 3 above)
Comment (sbegaudeau) [reply packages/diagrams/frontend/sirius-components-diagrams/src/renderer/dropNode/useDropNodeStyle.ts:45]: Same comment too :)

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/connector/useConnectorNodeStyle.tsx (hunk 1 above)
Comment (frouene) [reply packages/diagrams/frontend/sirius-components-diagrams/src/renderer/connector/useConnectorNodeStyle.tsx:52]: I didn't know the term, thanks !

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/node/FreeFormNode.tsx (hunk 4 above)
Comment (mcharfadi) [reply packages/diagrams/frontend/sirius-components-diagrams/src/renderer/node/FreeFormNode.tsx:85]: yes, thanks

**Review by frouene — APPROVED** [review]

**Pull request merged**